File: uploader/src/main/java/net/hearthstats/Updater.java
package net.hearthstats;

import net.hearthstats.config.Environment;
import net.hearthstats.log.Log;
import net.hearthstats.updater.api.GitHubReleases;
import net.hearthstats.updater.api.model.Release;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.awt.*;
import java.io.File;
import java.net.URI;


public final class Updater {
	private Updater() {} // never instantiated

  public static final String MANUAL_DOWNLOAD_URL = "http://hearthstats.net/uploader";

  private final static Logger debugLog = LoggerFactory.getLogger(Updater.class);

  private static Release cachedlatestRelease;


  public static Release getLatestRelease() {
    if (cachedlatestRelease == null) {
      try {
        debugLog.debug("Loading latest release information from GitHub");
        cachedlatestRelease = GitHubReleases.getLatestRelease(true);
        if (cachedlatestRelease == null) {
          Log.warn("Unable to check latest release of HearthStats Uploader");
        }
      } catch (Exception e) {
        Log.warn("Unable to check latest release of HearthStats Uploader due to error: " + e.getMessage(), e);
      }
    }
    return cachedlatestRelease;
  }


  public static String getAvailableVersion() {
    Release release = getLatestRelease();
    return release == null ? null : release.getVersion();
  }


  public static String getRecentChanges() {
    Release release = getLatestRelease();
    return release == null ? null : release.getBody();
  }


  public static void run(Environment environment, Release release) {
    Log.info("Extracting and running updater ...");

    String errorMessage = environment.performApplicationUpdate(release);

    if (errorMessage == null) {
      // There was no error, so the updater should be running now
      System.exit(0);

    } else {
      // There was an error
      Main.showMessageDialog(null, errorMessage + "\n\nYou will now be taken to the download page.");
      try {
        Desktop.getDesktop().browse(new URI(MANUAL_DOWNLOAD_URL));
      } catch (Exception e) {
        Main.showErrorDialog("Error launching browser with URL " + MANUAL_DOWNLOAD_URL, e);
      }
      Log.warn("Updater Error: " + errorMessage);
    }


  }


  public static void cleanUp() {
		removeFile(Config.getExtractionFolder() + "/updater.jar");
		removeFile(Config.getExtractionFolder() + "/update.zip");
	}


	private static void removeFile(String path) {
		File file = new File(path);
		if (file.isFile()) {
			file.delete();
		}
	}

}

File: uploader/src/main/scala/net/hearthstats/util/Updater.scala
package net.hearthstats.util

import net.hearthstats.config.Environment
import java.awt.Desktop
import java.net.URI
import net.hearthstats.updater.api.model.Release
import org.apache.commons.io.filefilter.WildcardFileFilter
import net.hearthstats.updater.api.GitHubReleases
import net.hearthstats.Main
import net.hearthstats.config.OS
import java.io.File
import net.hearthstats.ui.log.Log
import grizzled.slf4j.Logging
import java.io.FileFilter

class Updater(environment: Environment, uiLog: Log) extends Logging {

  val MANUAL_DOWNLOAD_URL = "http://hearthstats.net/uploader"

  private var cachedlatestRelease: Release = _

  def getLatestRelease(): Release = {
    if (cachedlatestRelease == null) {
      try {
        debug("Loading latest release information from GitHub")
        if (environment.os == OS.OSX) {
          cachedlatestRelease = GitHubReleases.getLatestReleaseForOSX
        } else if (environment.os == OS.WINDOWS) {
          cachedlatestRelease = GitHubReleases.getLatestReleaseForWindows
        }
        if (cachedlatestRelease == null) {
          uiLog.warn("Unable to check latest release of HearthStats Companion")
        }
      } catch {
        case e: Exception => uiLog.warn("Unable to check latest release of HearthStats Companion due to error: " +
          e.getMessage, e)
      }
    }
    cachedlatestRelease
  }

  def run(release: Release) {
    uiLog.info("Extracting and running updater ...")
    var errorMessage: String = null
    try {
      errorMessage = environment.performApplicationUpdate(release)
    } catch {
      case e: Exception => {
        warn("Unable to run updater", e)
        if (errorMessage == null) {
          errorMessage = e.getMessage
        }
      }
    }
    if (errorMessage == null) {
      System.exit(0)
    } else {
      Main.showMessageDialog(null, errorMessage + "\n\nYou will now be taken to the download page.")
      try {
        Desktop.getDesktop.browse(new URI(MANUAL_DOWNLOAD_URL))
      } catch {
        case e: Exception => Main.showErrorDialog("Error launching browser with URL " + MANUAL_DOWNLOAD_URL,
          e)
      }
      uiLog.warn("Updater Error: " + errorMessage)
    }
  }

  def cleanUp() {
    removeFile(environment.extractionFolder, "updater.jar")
    removeFile(environment.extractionFolder, "update-*.zip")
  }

  /**
   * Deletes a file or files in the given path.
   * @param path The path where the file or files are located.
   * @param filename The file or files to delete. Accepts wildcards * and ?.
   */
  private def removeFile(path: String, filename: String) {
    debug(s"Deleting $path/$filename")
    val dir = new File(path)
    val fileFilter: FileFilter = new WildcardFileFilter(filename)
    val files = dir.listFiles(fileFilter)
    for (file <- files) {
      info("Deleting file " + file.getAbsolutePath)
      file.delete()
    }
  }
}
